control-plane/gateways/service.go

```diff
@@ -4,26 +4,25 @@
 package gateways
 
 import (
+	meshv2beta1 "github.com/hashicorp/consul-k8s/control-plane/api/mesh/v2beta1"
 	corev1 "k8s.io/api/core/v1"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/apimachinery/pkg/util/intstr"
-
-	meshv2beta1 "github.com/hashicorp/consul-k8s/control-plane/api/mesh/v2beta1"
 )
 
-const port = int32(443)
-
 func (b *meshGatewayBuilder) Service() *corev1.Service {
 	var (
-		containerConfig *meshv2beta1.GatewayClassContainerConfig
 		portModifier    = int32(0)
+		port            = int32(443)
 		serviceType     = corev1.ServiceType("")
+		containerConfig *meshv2beta1.GatewayClassContainerConfig
 	)
 
 	if b.gcc != nil {
 		containerConfig = b.gcc.Spec.Deployment.Container
 		portModifier = containerConfig.PortModifier
 		serviceType = *b.gcc.Spec.Service.Type
+		port = b.gcc.Spec.Service.Port
```

Comment:
This is where you are getting that error you posted. The type of `Port` is `int32`. This will initialize by default to `0`. So you need an additional check when you determine the port. You'll need to check not only that `gcc` is not `nil`, but also that `b.gcc.Spec.Service.Port` is not `0`. 

What's happening in this case is you are setting `port` to `int32(443)`, then you are checking if `gcc` is not `nil`. It isn't, but `b.gcc.Spec.Service.Port == int32(0)` so you overwrite what you set `port` to. When the `Service` is created, you get an error because the port cannot be `0`.